crates/storage/provider/src/post_state.rs

```diff
@@ -0,0 +1,828 @@
+    /// [Change::StorageWiped] will also be present to mark the deletion of storage.
+    ///
+    /// If a change to an account satisfies the conditions for EIP-158, this change variant is also
+    /// applied instead of the change that would otherwise have happened.
+    AccountDestroyed {
+        /// The ID of the transition this change is a part of.
+        id: TransitionId,
+        /// The address of the destroyed account.
+        address: Address,
+        /// The account before it was destroyed.
+        old: Account,
+    },
+}
+
+impl Change {
+    /// Get the transition ID for the change
+    pub fn transition_id(&self) -> TransitionId {
+        match self {
+            Change::AccountChanged { id, .. } |
+            Change::AccountCreated { id, .. } |
+            Change::StorageChanged { id, .. } |
+            Change::StorageWiped { id, .. } |
+            Change::AccountDestroyed { id, .. } => *id,
+        }
+    }
+
+    /// Get the address of the account this change operates on.
+    pub fn address(&self) -> Address {
+        match self {
+            Change::AccountChanged { address, .. } |
+            Change::AccountCreated { address, .. } |
+            Change::StorageChanged { address, .. } |
+            Change::StorageWiped { address, .. } |
+            Change::AccountDestroyed { address, .. } => *address,
+        }
+    }
+
+    /// Set the transition ID of this change.
+    pub fn set_transition_id(&mut self, new_id: TransitionId) {
+        match self {
+            Change::AccountChanged { ref mut id, .. } |
+            Change::AccountCreated { ref mut id, .. } |
+            Change::StorageChanged { ref mut id, .. } |
+            Change::StorageWiped { ref mut id, .. } |
+            Change::AccountDestroyed { ref mut id, .. } => {
+                *id = new_id;
+            }
+        }
+    }
+}
+
+/// The state of accounts after execution of one or more transactions, including receipts and new
+/// bytecode.
+///
+/// The latest state can be found in `accounts`, `storage`, and `bytecode`. The receipts for the
+/// transactions that lead to these changes can be found in `receipts`, and each change leading to
+/// this state can be found in `changes`.
+///
+/// # Wiped Storage
+///
+/// The [Storage] type has a field, `wiped`, which denotes whether any of the values contained
+/// in storage are valid or not; if `wiped` is `true`, the storage for the account should be
+/// considered empty.
+///
+/// # Transitions
+///
+/// Each [Change] has an `id` field that marks what transition it is part of. Each transaction is
+/// its own transition, but there may be 0 or 1 transitions associated with the block.
+///
+/// The block level transition includes:
+///
+/// - Block rewards
+/// - Ommer rewards
+/// - Withdrawals
+/// - The irregular state change for the DAO hardfork
+///
+/// [PostState::finish_transition] should be called after every transaction, and after every block.
```

Comment:
Should or MUST? I had the thought that it might be a footgun if you forget to call it?